### PR TITLE
KPE-295 klepsydra ros2 plugin pendulum control stats

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp [hunk 1]
```diff
@@ -24,32 +10,31 @@
...
-#include "rclcpp/executor.hpp"
-#include "rclcpp/macros.hpp"
-#include "rclcpp/memory_strategies.hpp"
+#include <kpsr_ros2_executor/executor_factory.hpp>
+#include <kpsr_ros2_executor/streaming_executor.hpp>
 
 namespace pendulum_control
 {
-/// Instrumented executor that syncs Executor::spin functions with rttest_spin.
-class RttExecutor : public rclcpp::Executor
```
Comment (pablogh-2000) [inline pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: I think we can keep the name and the inheritance from the `Executor` interface

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp [hunk 2]
```diff
@@ -135,9 +124,10 @@ class RttExecutor : public rclcpp::Executor
   timespec start_time_;
 
 private:
-  RCLCPP_DISABLE_COPY(RttExecutor)
+  rclcpp::Executor::SharedPtr kpsrExecutor;
```
Comment (pablogh-2000) [inline pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Rename `kpsrExecutor` to `decorableExecutor`

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp [hunk 3]
```diff
@@ -24,32 +10,31 @@
...
   /// Constructor
   /**
    * Extends default Executor constructor
    */
-  RttExecutor(
+  RttKpsrExecutor(
     const rclcpp::ExecutorOptions & options = rclcpp::ExecutorOptions())
-  : rclcpp::Executor(options), running(false)
+  : running(false)
+  , kpsrExecutor(kpsr::ros2::ExecutorFactory::createExecutor(false, options))
```
Comment (pablogh-2000) [inline pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Do not create the executor here, but pass it as constructor argument, like the real decorator pattern:
https://sourcemaking.com/design_patterns/decorator

File: pendulum_control/src/pendulum_demo.cpp [hunk 4]
```diff
@@ -219,7 +219,7 @@ int main(int argc, char * argv[])
   options.memory_strategy = memory_strategy;
   // RttExecutor is a special single-threaded executor instrumented to calculate and record
   // real-time performance statistics.
-  auto executor = std::make_shared<pendulum_control::RttExecutor>(options);
+  auto executor = std::make_shared<pendulum_control::RttKpsrExecutor>(options);
```
Comment (pablogh-2000) [inline pendulum_control/src/pendulum_demo.cpp]: In the test, we can create the executor and pass it to the `RttExecutor`

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp (hunk 2 above)
Comment (ghost) [reply pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Done

Comment (ghost) [reply pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Done

File: pendulum_control/src/pendulum_demo.cpp (hunk 4 above)
Comment (ghost) [reply pendulum_control/src/pendulum_demo.cpp]: Done

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp (hunk 3 above)
Comment (ghost) [reply pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Done

File: pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp (hunk 1 above)
Comment (ghost) [reply pendulum_control/include/pendulum_control/rtt_kpsr_executor.hpp]: Done

File: pendulum_control/include/pendulum_control/rtt_executor.hpp [hunk 5]
```diff
@@ -103,12 +90,11 @@ class RttExecutor : public rclcpp::Executor
     // Cast the argument so that we can access the executor's state.
     RttExecutor * executor = static_cast<RttExecutor *>(arg);
     // If the input pointer was NULL or invalid, or if rclcpp has stopped, signal rttest to stop.
-    if (!executor || !rclcpp::ok()) {
```
Comment (pablogh-2000) [inline pendulum_control/include/pendulum_control/rtt_executor.hpp]: I think we can change the static method to receive no arguments. That would make it nicer for everyone, including the ROS Team. In this case, we need to split the PR in two parts. We will discuss this tomorrow.

Comment (ghost) [reply pendulum_control/include/pendulum_control/rtt_executor.hpp]: Done

**Review by mharshe — APPROVED** [review]

**Review by pablogh-2000 — APPROVED** [review]

**Review by mharshe — APPROVED** [review]

**Review by pablogh-2000 — APPROVED** [review]

File: pendulum_control/src/pendulum_demo.cpp [hunk 6]
```diff
@@ -129,12 +130,15 @@ int main(int argc, char * argv[])
   // message pool is determined by the number of threads (the maximum number of concurrent accesses
   // to the subscription).
   // Since this example is single-threaded, we choose a message pool size of 1 for each strategy.
+  const size_t messagePoolSize = 32;
```
Comment (pablogh-2000) [inline pendulum_control/src/pendulum_demo.cpp]: We still need that in our case we use this size?

Comment (ghost) [reply pendulum_control/src/pendulum_demo.cpp]: Done

**Review by pablogh-2000 — COMMENTED** [review]
Please rebase, test and if works, then merge.
Happy to review again if needed.

**Pull request merged**